lib/improver/convection.py

```diff
@@ -0,0 +1,322 @@
+                between adjacent grid squares along x and y, respectively.
+        """
+        diff_along_x_cube, diff_along_y_cube = (
+            DifferenceBetweenAdjacentGridSquares().process(cube))
+        # Compute the absolute values of the differences to ensure that
+        # negative differences are included.
+        diff_along_x_cube.data = np.absolute(diff_along_x_cube.data)
+        diff_along_y_cube.data = np.absolute(diff_along_y_cube.data)
+        cubelist = iris.cube.CubeList([diff_along_x_cube, diff_along_y_cube])
+        return cubelist
+
+    def iterate_over_threshold(self, cubelist, threshold):
+        """
+        Iterate over the application of thresholding to multiple cubes.
+
+        Args:
+            cubelist : Iris.cube.CubeList
+                Cubelist containing cubes to be thresholded.
+            threshold : float
+                The threshold that will be applied.
+
+        Returns:
+            cubes : Iris.cube.CubeList
+                Cubelist after thresholding each cube.
+        """
+        cubes = iris.cube.CubeList([])
+        for cube in cubelist:
+            threshold_cube = (
+                BasicThreshold(
+                    threshold, fuzzy_factor=self.fuzzy_factor,
+                    below_thresh_ok=self.below_thresh_ok
+                    ).process(cube.copy()))
+            cubes.append(threshold_cube)
+        return cubes
+
+    @staticmethod
+    def sum_differences_between_adjacent_grid_squares(cube, thresholded_cubes):
+        """
+        Put the differences back onto the original grid by summing together
+        the array with offsets. This covers the fact that the difference
+        cubes will result in output on a staggered grid compared with the
+        input cube.
+
+        Args:
+            cube : Iris.cube.Cube
+                The cube with the original grid.
+            thresholded_cubes : Iris.cube.CubeList
+                Cubelist containing differences between adjacent grid squares
+                along x and differences between adjacent grid squares along y,
+                which have been thresholded.
+
+        Args:
+            cube_on_orig_grid : Iris.cube.Cube
+                Cube on the original grid with the values from the thresholded
+                adjacent grid square difference cubes inserted. The resulting
+                values have been restricted to be between 0 and 1.
+        """
+        threshold_cube_x, threshold_cube_y = thresholded_cubes
+        cube_on_orig_grid = cube.copy()
+        cube_on_orig_grid.data = np.zeros(cube_on_orig_grid.shape)
+        cube_on_orig_grid.data[..., :-1, :] += threshold_cube_y.data
+        cube_on_orig_grid.data[..., 1:, :] += threshold_cube_y.data
+        cube_on_orig_grid.data[..., :, :-1] += threshold_cube_x.data
+        cube_on_orig_grid.data[..., :, 1:] += threshold_cube_x.data
+        cube_on_orig_grid.data = np.clip(cube_on_orig_grid.data, 0.0, 1.0)
```

Comment:
Is a value of 4 (1+1+1+1) equivalent to a value of 1 (0.25+0.25+0.25+0.25)? It will be after the clipping. If it's not equivalent, then we need to normalise rather than clip.